onnxruntime/core/session/onnxruntime_c_api_ep.cc

```diff
@@ -0,0 +1,1044 @@
+ *   1. if the given node name can't be found in the graph, add an new "EP Context" node to the existing graph
+ *   2. if the node is already existed, update the node attributes only
+ */
+ORT_API_STATUS_IMPL(OrtGraphApis::OrtGraph_CreateOrUpdateEpCtxGraph,
+                    const OrtGraphViewer* graph,
+                    const char* node_name,
+                    const int64_t main_context,
+                    const int64_t embed_mode,
+                    const char* cache_path,
+                    char* cache_data,
+                    size_t size,
+                    const char* const* extra_attr_keys,
+                    const char* const* extra_attr_values,
+                    size_t extra_attr_num,
+                    _Outptr_ OrtGraphViewer** ep_context_graph) {
+
+  const std::string EPCONTEXT_OP = "EPContext";
+  const std::string MAIN_CONTEXT = "main_context";
+  const std::string EMBED_MODE = "embed_mode";
+  const std::string EP_CACHE_CONTEXT = "ep_cache_context";
+  const std::string ONNX_MODEL_FILENAME = "onnx_model_filename";
+  const std::string EPCONTEXT_OP_DOMAIN = "com.microsoft";
+  const std::string EPCONTEXT_WARNING =
+    "It's suggested to set the ORT graph optimization level to 0 and  \
+                                              make \"embed_mode\" to 0 (\"ep_cache_context\" is the cache path)\
+                                              for the best model loading time";
+
+  const ::onnxruntime::GraphViewer* graph_viewer = reinterpret_cast<const ::onnxruntime::GraphViewer*>(graph);
+  ::onnxruntime::Graph* graph_build;
+
+  if (!graph_viewer && !(*ep_context_graph)) return nullptr;
+
+  std::unordered_map<std::string, std::string> attr_keys_values;
+  for (size_t i = 0; i < extra_attr_num; i++) {
+    attr_keys_values[extra_attr_keys[i]] = extra_attr_values[i];
+  }
+
+  // Create a new graph or use the existing one
+  if (*ep_context_graph == nullptr) {
+    Model* model_build = new Model (graph_viewer->Name(), true, ModelMetaData(), PathString(),
+#if !defined(ORT_MINIMAL_BUILD)
+                                   IOnnxRuntimeOpSchemaRegistryList({graph_viewer->GetSchemaRegistry()}), graph_viewer->DomainToVersionMap(),
+#else
+                                   IOnnxRuntimeOpSchemaRegistryList(), graph_viewer->DomainToVersionMap(),
+#endif  // ORT_MINIMAL_BUILD
+                                   std::vector<ONNX_NAMESPACE::FunctionProto>(), graph_viewer->GetGraph().GetLogger());
+    graph_build = &(model_build->MainGraph());
+    auto graph_build_viewer = std::make_unique<GraphViewer>(*graph_build);
+    *ep_context_graph = reinterpret_cast<OrtGraphViewer*>(graph_build_viewer.release());
+  } else {
+    ::onnxruntime::GraphViewer* content_graph_viewer = reinterpret_cast<::onnxruntime::GraphViewer*>(*ep_context_graph);
+    graph_build = const_cast<::onnxruntime::Graph*>(&(content_graph_viewer->GetGraph()));
```

Comment:
It seems using `const_cast()` here might result in 
"Modifying a const object through a non-const access path and referring to a volatile object through a non-volatile [glvalue](https://en.cppreference.com/w/cpp/language/value_category#glvalue) results in undefined behavior." - https://en.cppreference.com/w/cpp/language/const_cast
because later this function calls graph_build->GetOrCreateNodeArg which might modify the graph instance.

Reply:
Thanks for the comment. Rolled back the change